agent/eventhandler/task_handler.go

```diff
@@ -15,244 +15,323 @@ package eventhandler
+	state                dockerstate.TaskEngineState
+	client               api.ECSClient
+	ctx                  context.Context
+}
+
+// taskSendableEvents is used to group all events for a task
+type taskSendableEvents struct {
+	// events is a list of *sendableEvents
+	events *list.List
+	// sending will check whether the list is already being handlerd
+	sending bool
+	//eventsListLock locks both the list and sending bool
+	lock sync.Mutex
+	// createdAt is a timestamp for when the event list was created
+	createdAt time.Time
+	// taskARN is the task arn that the event list is associated
+	taskARN string
 }
 
 // NewTaskHandler returns a pointer to TaskHandler
-func NewTaskHandler(stateManager statemanager.Saver) *TaskHandler {
-	return &TaskHandler{
-		tasksToEvents:          make(map[string]*eventList),
+func NewTaskHandler(ctx context.Context,
+	stateManager statemanager.Saver,
+	state dockerstate.TaskEngineState,
+	client api.ECSClient) *TaskHandler {
+	// Create a handler and start the periodic event drain loop
+	taskHandler := &TaskHandler{
+		ctx:                    ctx,
+		tasksToEvents:          make(map[string]*taskSendableEvents),
 		submitSemaphore:        utils.NewSemaphore(concurrentEventCalls),
 		tasksToContainerStates: make(map[string][]api.ContainerStateChange),
 		stateSaver:             stateManager,
+		state:                  state,
+		client:                 client,
+		drainEventsFrequency:   drainEventsFrequency,
 	}
+	go taskHandler.startDrainEventsTicker()
+
+	return taskHandler
 }
 
 // AddStateChangeEvent queues up a state change for sending using the given client.
 func (handler *TaskHandler) AddStateChangeEvent(change statechange.Event, client api.ECSClient) error {
+	handler.lock.Lock()
+	defer handler.lock.Unlock()
+
 	switch change.GetEventType() {
 	case statechange.TaskEvent:
 		event, ok := change.(api.TaskStateChange)
 		if !ok {
 			return errors.New("eventhandler: unable to get task event from state change event")
 		}
-		handler.flushBatch(&event)
-		handler.addEvent(newSendableTaskEvent(event), client)
+		handler.flushBatchUnsafe(&event, client)
 		return nil
 
 	case statechange.ContainerEvent:
 		event, ok := change.(api.ContainerStateChange)
 		if !ok {
 			return errors.New("eventhandler: unable to get container event from state change event")
 		}
-		handler.batchContainerEvent(event)
+		handler.batchContainerEventUnsafe(event)
 		return nil
 
 	default:
 		return errors.New("eventhandler: unable to determine event type from state change event")
 	}
 }
 
-// batchContainerEvent collects container state change events for a given task arn
-func (handler *TaskHandler) batchContainerEvent(event api.ContainerStateChange) {
-	handler.taskHandlerLock.Lock()
-	defer handler.taskHandlerLock.Unlock()
+// startDrainEventsTicker starts a ticker that periodically drains the events queue
+// by submitting state change events to the ECS backend
+func (handler *TaskHandler) startDrainEventsTicker() {
```

Comment:
nit: I'd love to move this function before `AddStateChangeEvent`.

Reply:
I wanted the exported functions to be at the top of the file as that's a reasonable way to organize the code.